### PR TITLE
Turn IGatewayListProvider.GetGateways() into a Task-based method

### PR DESCRIPTION
**Pull request by dVakulen**
Addresses #927

### CODE REVIEW

File: src/Orleans/Async/TaskExtensions.cs [hunk 1]
```diff
@@ -130,11 +131,13 @@ internal static T WaitForResultWithThrow<T>(this Task<T> task, TimeSpan timeout)
                 return;
             }
 
-            await Task.WhenAny(taskToComplete, Task.Delay(timeout));
+            var timeoutCancellationTokenSource = new CancellationTokenSource();
```
Comment (sergeybykov) [inline src/Orleans/Async/TaskExtensions.cs]: Why this change? Seems unrelated to the rest.

Comment (dVakulen) [inline src/Orleans/Async/TaskExtensions.cs]: Should I fill new issue and make another PR? [Asynchronously wait for Task<T> to complete with timeout](http://stackoverflow.com/questions/4238345/asynchronously-wait-for-taskt-to-complete-with-timeout#comment27036701_11191070)

Comment (sergeybykov) [inline src/Orleans/Async/TaskExtensions.cs]: I think a separate PR will be better. Thanks.

**Comment by sergeybykov** [comment]
Testing it.

**Comment by sergeybykov** [comment]
Well done. Are we a running out of CoreCLR workitems? :-)

**Comment by gabikliot** [comment]
No, we are not. There is the second list. Plus, just run analyzer again and find more items to port.
